packages/next/src/build/webpack/loaders/next-app-loader.ts

```diff
@@ -302,9 +304,18 @@ async function createTreeCodeFromPath(
 
       const parallelSegmentPath = subSegmentPath.join('/')
 
+      const fileTypesToResolve: (FileTypeValues | 'page')[] =
+        Object.values(FILE_TYPES)
+
+      // Special case for exports in parent pages
+      const isSegmentDynamic = /^\[.*\]$/.test(normalizedParallelSegments[0])
+      const isFinalSegment =
+        normalizedParallelSegments[0] === page.split('/').at(-2)
```

Comment:
Had to remind myself why this would be `-2` which was a bit confusing. Perhaps we can first run page through [`normalizeAppPath`](https://github.com/vercel/next.js/blob/canary/packages/next/src/shared/lib/router/utils/app-paths.ts#L23) -- that'll give us something like `/gen-params-top-down-page/[lang]/[slug]`.  Then we can simplify this check to be `normalizedParallelSegments[0] === normalizedAppPath.at(-1)`